{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "8f6e3697",
   "metadata": {},
   "outputs": [],
   "source": [
    "import cobra\n",
    "import csv\n",
    "\n",
    "model = cobra.io.load_json_model('e_coli_core.json')\n",
    "\n",
    "filename = 'e_coli_core_expression.csv'\n",
    "estimated_maximal_activity = {}\n",
    "\n",
    "with open(filename, mode='r', newline='') as file:\n",
    "    reader = csv.reader(file)   \n",
    "    for row in reader:\n",
    "        \n",
    "        try:\n",
    "            num = float(row[1])\n",
    "            estimated_maximal_activity[row[0]] = num\n",
    "        except ValueError:\n",
    "            continue\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "972db4ab",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ATPM reaction unchanged\n",
      "EX_glc__D_e bounds changed to high absolute default bound\n"
     ]
    }
   ],
   "source": [
    "high_absolute_default_bound = 1000\n",
    "\n",
    "for reaction in model.reactions:\n",
    "    \n",
    "    if (reaction.id==\"ATPM\"):\n",
    "        print(\"ATPM reaction unchanged\")\n",
    "        continue\n",
    "    \n",
    "    if (reaction.id==\"EX_glc__D_e\"):\n",
    "        print(\"EX_glc__D_e bounds changed to high absolute default bound\")\n",
    "        reaction.lower_bound = -high_absolute_default_bound\n",
    "        reaction.upper_bound = high_absolute_default_bound\n",
    "        continue\n",
    "    \n",
    "    try:\n",
    "        value = estimated_maximal_activity[reaction.id]\n",
    "    except KeyError:\n",
    "        continue\n",
    "    \n",
    "    if (reaction.reversibility == True):\n",
    "        reaction.lower_bound = -value\n",
    "        reaction.upper_bound = value\n",
    "        \n",
    "    elif (reaction.reversibility == False):\n",
    "        reaction.lower_bound = 0.0\n",
    "        reaction.upper_bound = value"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "d3e82917",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PFK: UB:12.12, LB:0.0\n",
      "PFL: UB:1.0, LB:0.0\n",
      "PGI: UB:13.12, LB:-13.12\n",
      "PGK: UB:23.13, LB:-23.13\n",
      "PGL: UB:8.12, LB:0.0\n",
      "ACALD: UB:1.16, LB:-1.16\n",
      "AKGt2r: UB:3.1, LB:-3.1\n",
      "PGM: UB:20.01, LB:-20.01\n",
      "PIt2r: UB:6.03, LB:-6.03\n",
      "ALCD2x: UB:9.01, LB:-9.01\n",
      "ACALDt: UB:2.29, LB:-2.29\n",
      "ACKr: UB:1.19, LB:-1.19\n",
      "PPC: UB:2.56, LB:0.0\n",
      "ACONTa: UB:25.35, LB:-25.35\n",
      "ACONTb: UB:25.35, LB:-25.35\n",
      "ATPM: UB:1000.0, LB:8.39\n",
      "PPCK: UB:25.23, LB:0.0\n",
      "ACt2r: UB:3.23, LB:-3.23\n",
      "PPS: UB:2.5, LB:0.0\n",
      "ADK1: UB:30.57, LB:-30.57\n",
      "AKGDH: UB:24.35, LB:0.0\n",
      "ATPS4r: UB:60.5, LB:-60.5\n",
      "PTAr: UB:4.47, LB:-4.47\n",
      "PYK: UB:26.78, LB:0.0\n",
      "BIOMASS_Ecoli_core_w_GAM: UB:1000.0, LB:0.0\n",
      "PYRt2: UB:1.26, LB:-1.26\n",
      "CO2t: UB:30.45, LB:-30.45\n",
      "RPE: UB:5.67, LB:-5.67\n",
      "CS: UB:20.56, LB:0.0\n",
      "RPI: UB:5.56, LB:-5.56\n",
      "SUCCt2_2: UB:2.36, LB:0.0\n",
      "CYTBD: UB:40.56, LB:0.0\n",
      "D_LACt2: UB:4.56, LB:-4.56\n",
      "ENO: UB:28.78, LB:-28.78\n",
      "SUCCt3: UB:6.67, LB:0.0\n",
      "ETOHt2r: UB:2.34, LB:-2.34\n",
      "SUCDi: UB:24.34, LB:0.0\n",
      "SUCOAS: UB:20.6, LB:-20.6\n",
      "TALA: UB:4.45, LB:-4.45\n",
      "THD2: UB:4.5, LB:0.0\n",
      "TKT1: UB:3.34, LB:-3.34\n",
      "TKT2: UB:3.34, LB:-3.34\n",
      "TPI: UB:45.56, LB:-45.56\n",
      "EX_ac_e: UB:1000.0, LB:0.0\n",
      "EX_acald_e: UB:1000.0, LB:0.0\n",
      "EX_akg_e: UB:1000.0, LB:0.0\n",
      "EX_co2_e: UB:1000.0, LB:-1000.0\n",
      "EX_etoh_e: UB:1000.0, LB:0.0\n",
      "EX_for_e: UB:1000.0, LB:0.0\n",
      "EX_fru_e: UB:1000.0, LB:0.0\n",
      "EX_fum_e: UB:1000.0, LB:0.0\n",
      "EX_glc__D_e: UB:1000, LB:-1000\n",
      "EX_gln__L_e: UB:1000.0, LB:0.0\n",
      "EX_glu__L_e: UB:1000.0, LB:0.0\n",
      "EX_h_e: UB:1000.0, LB:-1000.0\n",
      "EX_h2o_e: UB:1000.0, LB:-1000.0\n",
      "EX_lac__D_e: UB:1000.0, LB:0.0\n",
      "EX_mal__L_e: UB:1000.0, LB:0.0\n",
      "EX_nh4_e: UB:1000.0, LB:-1000.0\n",
      "EX_o2_e: UB:1000.0, LB:-1000.0\n",
      "EX_pi_e: UB:1000.0, LB:-1000.0\n",
      "EX_pyr_e: UB:1000.0, LB:0.0\n",
      "EX_succ_e: UB:1000.0, LB:0.0\n",
      "FBA: UB:24.45, LB:-24.45\n",
      "FBP: UB:2.35, LB:0.0\n",
      "FORt2: UB:1.28, LB:0.0\n",
      "FORt: UB:2.34, LB:0.0\n",
      "FRD7: UB:20.07, LB:0.0\n",
      "FRUpts2: UB:10.08, LB:0.0\n",
      "FUM: UB:24.35, LB:-24.35\n",
      "FUMt2_2: UB:2.45, LB:0.0\n",
      "G6PDH2r: UB:5.68, LB:-5.68\n",
      "GAPD: UB:26.13, LB:-26.13\n",
      "GLCpts: UB:16.34, LB:0.0\n",
      "GLNS: UB:3.45, LB:0.0\n",
      "GLNabc: UB:0.25, LB:0.0\n",
      "GLUDy: UB:7.35, LB:-7.35\n",
      "GLUN: UB:6.25, LB:0.0\n",
      "GLUSy: UB:7.35, LB:0.0\n",
      "GLUt2r: UB:4.12, LB:-4.12\n",
      "GND: UB:6.72, LB:0.0\n",
      "H2Ot: UB:20.06, LB:-20.06\n",
      "ICDHyr: UB:15.06, LB:-15.06\n",
      "ICL: UB:14.07, LB:0.0\n",
      "LDH_D: UB:2.07, LB:-2.07\n",
      "MALS: UB:9.12, LB:0.0\n",
      "MALt2_2: UB:8.87, LB:0.0\n",
      "MDH: UB:4.26, LB:-4.26\n",
      "ME1: UB:3.24, LB:0.0\n",
      "ME2: UB:3.34, LB:0.0\n",
      "NADH16: UB:20.26, LB:0.0\n",
      "NADTRHD: UB:1.26, LB:0.0\n",
      "NH4t: UB:3.45, LB:-3.45\n",
      "O2t: UB:40.67, LB:-40.67\n",
      "PDH: UB:32.32, LB:0.0\n"
     ]
    }
   ],
   "source": [
    "for reaction in model.reactions:\n",
    "    print(f'{reaction.id}: UB:{reaction.upper_bound}, LB:{reaction.lower_bound}')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "20eb692f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Flux Variability Results:\n",
      "                               minimum       maximum\n",
      "PFK                       7.314456e+00  7.314456e+00\n",
      "PFL                      -3.165524e-14  0.000000e+00\n",
      "PGI                       7.376438e+00  7.376438e+00\n",
      "PGK                      -1.446649e+01 -1.446649e+01\n",
      "PGL                       4.751035e-01  4.751035e-01\n",
      "ACALD                    -1.160000e+00 -1.160000e+00\n",
      "AKGt2r                   -1.098154e-14  0.000000e+00\n",
      "PGM                      -1.374906e+01 -1.374906e+01\n",
      "PIt2r                     1.764199e+00  1.764199e+00\n",
      "ALCD2x                   -1.160000e+00 -1.160000e+00\n",
      "ACALDt                   -1.446793e-14 -4.784464e-15\n",
      "ACKr                     -1.190000e+00 -1.190000e+00\n",
      "PPC                       1.374260e+00  1.374260e+00\n",
      "ACONTa                    3.289982e+00  3.289982e+00\n",
      "ACONTb                    3.289982e+00  3.289982e+00\n",
      "ATPM                      8.390000e+00  8.390000e+00\n",
      "PPCK                      0.000000e+00  8.135426e-15\n",
      "ACt2r                    -1.190000e+00 -1.190000e+00\n",
      "PPS                       0.000000e+00  1.211304e-14\n",
      "ADK1                      0.000000e+00  1.227109e-14\n",
      "AKGDH                     2.772573e+00  2.772573e+00\n",
      "ATPS4r                    2.190519e+01  2.190519e+01\n",
      "PTAr                      1.190000e+00  1.190000e+00\n",
      "PYK                       4.175997e+00  4.175997e+00\n",
      "BIOMASS_Ecoli_core_w_GAM  4.795714e-01  4.795714e-01\n",
      "PYRt2                    -1.260000e+00 -1.260000e+00\n",
      "CO2t                     -1.260072e+01 -1.260072e+01\n",
      "RPE                      -2.798030e-02 -2.798030e-02\n",
      "CS                        3.289982e+00  3.289982e+00\n",
      "RPI                      -5.030838e-01 -5.030838e-01\n",
      "SUCCt2_2                  0.000000e+00  1.647232e-14\n",
      "CYTBD                     2.303257e+01  2.303257e+01\n",
      "D_LACt2                  -2.070000e+00 -2.070000e+00\n",
      "ENO                       1.374906e+01  1.374906e+01\n",
      "SUCCt3                    0.000000e+00  1.689190e-14\n",
      "ETOHt2r                  -1.160000e+00 -1.160000e+00\n",
      "SUCDi                     2.772573e+00  2.284257e+01\n",
      "SUCOAS                   -2.772573e+00 -2.772573e+00\n",
      "TALA                      7.257249e-02  7.257249e-02\n",
      "THD2                      4.500000e+00  4.500000e+00\n",
      "TKT1                      7.257249e-02  7.257249e-02\n",
      "TKT2                     -1.005528e-01 -1.005528e-01\n",
      "TPI                       7.314456e+00  7.314456e+00\n",
      "EX_ac_e                   1.190000e+00  1.190000e+00\n",
      "EX_acald_e                0.000000e+00  1.411913e-14\n",
      "EX_akg_e                  0.000000e+00  1.349424e-14\n",
      "EX_co2_e                  1.260072e+01  1.260072e+01\n",
      "EX_etoh_e                 1.160000e+00  1.160000e+00\n",
      "EX_for_e                  0.000000e+00  0.000000e+00\n",
      "EX_fru_e                  0.000000e+00  0.000000e+00\n",
      "EX_fum_e                  0.000000e+00  0.000000e+00\n",
      "EX_glc__D_e              -7.949853e+00 -7.949853e+00\n",
      "EX_gln__L_e               0.000000e+00  6.924752e-15\n",
      "EX_glu__L_e               0.000000e+00  6.844164e-15\n",
      "EX_h_e                    1.414020e+01  1.414020e+01\n",
      "EX_h2o_e                  1.619411e+01  1.619411e+01\n",
      "EX_lac__D_e               2.070000e+00  2.070000e+00\n",
      "EX_mal__L_e               0.000000e+00  0.000000e+00\n",
      "EX_nh4_e                 -2.615007e+00 -2.615007e+00\n",
      "EX_o2_e                  -1.151629e+01 -1.151629e+01\n",
      "EX_pi_e                  -1.764199e+00 -1.764199e+00\n",
      "EX_pyr_e                  1.260000e+00  1.260000e+00\n",
      "EX_succ_e                 0.000000e+00  4.347736e-15\n",
      "FBA                       7.314456e+00  7.314456e+00\n",
      "FBP                       0.000000e+00  1.235424e-14\n",
      "FORt2                     0.000000e+00  2.940110e-14\n",
      "FORt                      0.000000e+00  4.958458e-14\n",
      "FRD7                      0.000000e+00  2.007000e+01\n",
      "FRUpts2                   0.000000e+00  0.000000e+00\n",
      "FUM                       2.772573e+00  2.772573e+00\n",
      "FUMt2_2                   0.000000e+00  0.000000e+00\n",
      "G6PDH2r                   4.751035e-01  4.751035e-01\n",
      "GAPD                      1.446649e+01  1.446649e+01\n",
      "GLCpts                    7.949853e+00  7.949853e+00\n",
      "GLNS                      1.226264e-01  1.226264e-01\n",
      "GLNabc                   -6.924752e-15  0.000000e+00\n",
      "GLUDy                    -2.492381e+00 -2.492381e+00\n",
      "GLUN                      0.000000e+00  1.117776e-14\n",
      "GLUSy                     0.000000e+00  1.609886e-14\n",
      "GLUt2r                   -4.830818e-15  0.000000e+00\n",
      "GND                       4.751035e-01  4.751035e-01\n",
      "H2Ot                     -1.619411e+01 -1.619411e+01\n",
      "ICDHyr                    3.289982e+00  3.289982e+00\n",
      "ICL                       0.000000e+00  1.938086e-14\n",
      "LDH_D                    -2.070000e+00 -2.070000e+00\n",
      "MALS                      0.000000e+00  2.571132e-14\n",
      "MALt2_2                   0.000000e+00  0.000000e+00\n",
      "MDH                       2.772573e+00  2.772573e+00\n",
      "ME1                       0.000000e+00  6.322335e-15\n",
      "ME2                       0.000000e+00  3.417029e-14\n",
      "NADH16                    2.026000e+01  2.026000e+01\n",
      "NADTRHD                   0.000000e+00  1.976678e-14\n",
      "NH4t                      2.615007e+00  2.615007e+00\n",
      "O2t                       1.151629e+01  1.151629e+01\n",
      "PDH                       7.437320e+00  7.437320e+00\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "from cobra.flux_analysis import flux_variability_analysis\n",
    "\n",
    "pd.set_option('display.max_rows', None)\n",
    "fva_results = flux_variability_analysis(model, model.reactions, fraction_of_optimum=1.0)\n",
    "\n",
    "print(\"Flux Variability Results:\")\n",
    "print(fva_results)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "34ddaca9",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of reactions limited by the network: 30\n",
      "\n",
      "List of constrained reaction IDs:\n",
      "['PFK', 'PGI', 'PGL', 'PIt2r', 'PPC', 'ACONTa', 'ACONTb', 'AKGDH', 'ATPS4r', 'PTAr', 'PYK', 'CS', 'CYTBD', 'ENO', 'SUCDi', 'TALA', 'TKT1', 'TPI', 'FBA', 'FUM', 'G6PDH2r', 'GAPD', 'GLCpts', 'GLNS', 'GND', 'ICDHyr', 'MDH', 'NH4t', 'O2t', 'PDH']\n"
     ]
    }
   ],
   "source": [
    "\n",
    "constrained_reactions = []\n",
    "tolerance = 1e-9\n",
    "\n",
    "\n",
    "for reaction in model.reactions:\n",
    "    if reaction.id in estimated_maximal_activity:\n",
    "        fva_max_flux = fva_results.loc[reaction.id, 'maximum']\n",
    "        reaction_upper_bound = reaction.upper_bound\n",
    "        \n",
    "        if (reaction.id != 'FORt' and\n",
    "            fva_max_flux > tolerance and  \n",
    "            fva_max_flux < reaction_upper_bound - tolerance): \n",
    "            \n",
    "            constrained_reactions.append(reaction.id)\n",
    "\n",
    "print(f\"Number of reactions limited by the network: {len(constrained_reactions)}\")\n",
    "print(\"\\nList of constrained reaction IDs:\")\n",
    "print(constrained_reactions)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "29d0ec78",
   "metadata": {},
   "source": [
    "Reaction behaviour explanation:\n",
    "There are a total of 30 reactions that the permissible flux range in the forward direction is non-zero, but they come out less than the gene expression imposed upper flux bound. This is likely to be explained by the fact that a reactions maximum possible speed is limited by two things:\n",
    "Their upper bond, so local capacity and the FVA result so in a sense their systemic throughput. That is the true maximum rate the reaction can occur given all other limitations in the network. We can say that these 30 reactions are not the bottleneck in the system but rather the network itself is the problem. Likely real bottlenecks would be upstream limits: for example the rate of glucose intake (GLCpts at 7,95) this limits any reaction downstream of that. Also the reactions themselves consuming its products are at their own maximum capacity. "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ebfeb241",
   "metadata": {},
   "source": [
    "Based on the flux variablity results, there are a total of 40 reactions that have a positive minimum flux. For these reactions, it is essential that a reaction happens to fulfill the specified cellular objective, in this case it is maximum biomass production. They must carry a forward flux, thus cannot be turned off or reversed. For some of these reactions this is the case because they are the only link in the pathway that can produce a critical component in that pathway like a specific amino acid for example. Also the entire biomass reaction itself must be in a positive flux for the model to simulate and maintain growth. In some way, those specific reactions are essential to the cells metabolism. "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dda4455e",
   "metadata": {},
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "BioSim",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
